### ISSUE TITLE
Noto Sans (combining) diacritical marks misaligned when coloured or in bold

### ISSUE THREAD

**Issue by techmagus** [notofonts/latin-greek-cyrillic#399]
* `/fonts/NotoSans/full/otf/NotoSans-Regular.otf`
* `/fonts/NotoSans/full/otf/NotoSans-Regular.ttf`
* From: https://github.com/notofonts/notofonts.github.io
* Date: 2023-01-31
* Version: 2.010
* OS: Pop!_OS 24.01 LTS
* Mozilla Firefox 110.0b9
* Microsoft Edge Version 111.0.1652.0 (Chromium-based)
* Naver Whale Version 3.19.166.16 (Chromium-based)

## Issue

When adding a colour to diacritical marks (or making it bold), the marks are then misaligned.

Normally, a software will replace a character with a diacritical mark with a precomposed Unicode equivalent if it exists. However, when enclosing a diacritical mark within `<span></span>`, it stays as-is instead of being replaced. This, in turn, makes the diacritical mark misaligned.

This was tested in DejaVu Sans and FreeSans (latest versions) as well. In DejaVu Sans it works fine; while in FreeSans the misalignment is slight. If the `font-family` of the diacritical mark is changed to DejaVu Sans while retaining the rest of the text as Noto Sans, it looks better (see screenshots and test-repo below for more).

### Character data

10 combining diacritical marks tested:

* U+0300 combining grave
* U+0303 combining tildə
* U+0360 combining double tildə
* U+0302 combining circumflex
* U+0331 combining macron below
* U+035F combining double macron below
* U+0304 combining macron
* U+035E combining double macron
* U+0301 combining acute
* U+0308 combining dieresis

**Mozilla Firefox 110.0b9**
![Screenshot from 2023-02-18 02-36-56](https://user-images.githubusercontent.com/1079205/219758972-7046886d-9a08-4aaa-a1f0-f501f6bce28f.png)

**Microsoft Edge Version 111.0.1652.0 (Chromium-based)**
![Screenshot from 2023-02-18 02-38-27](https://user-images.githubusercontent.com/1079205/219758990-bc4f11c8-049c-47a5-a1cb-d20e77a32fe6.png)

**Naver Whale Version 3.19.166.16 (Chromium-based)**
![Screenshot from 2023-02-18 02-38-45](https://user-images.githubusercontent.com/1079205/219759013-ee17cb6f-127f-475e-b92d-f5f0879a9c4e.png)

## Test file

A test file is available in my test-repo [https://github.com/techmagus/test-repo/tree/noto-diacriticals](https://github.com/techmagus/test-repo/tree/noto-diacriticals) under the `noto-diacriticals` branch.

**Comment by simoncozens** [comment]
I'm not sure this is something that can be fixed, except in the rendering engine. What should happen is that the diacritic and base character are anchored together using mark-to-base positioning, but when they are set in different colours, the base and mark are placed in different text runs so that mark-to-base positioning does not apply. 

The way that FreeSans and DejaVu are doing this is by having a large negative sidebearing on the left side of the diacritic, so that it gets positioned *somewhere* over the base character; you get the same absolute positioning whether the base character is wide or narrow. It *kind of* works in situations when all your base characters are approximately the same width, but it's a hack.

Basically this is just a limitation of how OpenType works.

**Comment by techmagus** [comment]
> I'm not sure this is something that can be fixed, except in the rendering engine. What should happen is that the diacritic and base character are anchored together using mark-to-base positioning, but when they are set in different colours, the base and mark are placed in different text runs so that mark-to-base positioning does not apply.
> 
> The way that FreeSans and DejaVu are doing this is by having a large negative sidebearing on the left side of the diacritic, so that it gets positioned _somewhere_ over the base character; you get the same absolute positioning whether the base character is wide or narrow. It _kind of_ works in situations when all your base characters are approximately the same width, but it's a hack.
> 
> Basically this is just a limitation of how OpenType works.

Ahh.

Thank you for the explanation, makes perfect sense. Deeply appreciated!

**Comment by khaledhosny** [comment]
Color shouldn’t break mark attachment, at least in Firefox. Bold will break it as it is a different font like Simon said.

For Firefox, you also need to force-off use of precomposed forms (I used U+034F between the base and the mark). Chrome, will not even apply color to combining mark alone, even when decomposed (e.g. x̂ which has no composed form).
<img width="93" alt="image" src="https://user-images.githubusercontent.com/93914/219937050-a0262182-368e-44ae-b3df-574f66b8432d.png">

**Comment by techmagus** [comment]
> Color shouldn’t break mark attachment, at least in Firefox. Bold will break it as it is a different font like Simon said.
> 
> (first without color, second with color, third with color and U+034F).

Actually, even if it is not in bold, it also breaks. Here are screenshots wherein the font of the diacritics is Noto Sans. It's the same in Firefox, Edge, and Whale.

![Screenshot from 2023-02-21 11-20-12](https://user-images.githubusercontent.com/1079205/220239597-122c888e-e620-4b97-b85d-ad499c2057de.png)
![Screenshot from 2023-02-21 11-20-28](https://user-images.githubusercontent.com/1079205/220239603-bbd24e22-ead8-47ba-bb60-1675eabd5de7.png)
![Screenshot from 2023-02-21 11-21-03](https://user-images.githubusercontent.com/1079205/220239611-6cf6928c-0df9-4f40-8a3d-4e994fc6f3dc.png)
![Screenshot from 2023-02-21 11-22-28](https://user-images.githubusercontent.com/1079205/220239648-597788a2-0b19-4776-b9f9-78d9b229bb00.png)

It's possible to break the precomposed character without using U+034F, in Firefox and Chromium-based browsers.

In any case, it's not important in the grand scheme of things. I just encountered it while building on a project, and filed a report about it because DejaVu Sans doesn't seem to be affected (or at least, not noticeable). And I was approaching it from a regular user vantage point. A workaround I can settle with is to just use DejaVu Sans for specific diacritical marks.

But, yeah, as what simoncozens said, a limitation on how OpenType works.

^_^

**Comment by khaledhosny** [comment]
That is because you are both changing the color and making the diacritics use bold font. The bold bit can’t be made to work, but changing the color alone is possible in Firefox (if decomposed form is used). Chrome will not change the color even if decomposed form is forced as it treats grapheme clusters as one unit for styling even if they are composed of multiple glyphs. WebKit is even worse, inline styling always breaks text layout even if the same font is used.

**Comment by khaledhosny** [comment]
Applying this to your HTML file:
```diff
diff --git a/diacritical-mark-test-page.html b/diacritical-mark-test-page.html
index 001da55..097a5cc 100644
--- a/diacritical-mark-test-page.html
+++ b/diacritical-mark-test-page.html
@@ -60,8 +60,8 @@
         color: hsla(0deg, 100%, 50%, 1);
         color: hwb(0deg 0% 0% / 100%);
       }
-      span.text-guide.diacritical-mark {
-        font-weight: bold;
+      span.text-guide.diacritical-mark::before {
+        content: "\034F"
       }
       span.text-guide.diacritical-mark.deja-font {
         color: blue;
```

Gives me (in Firefox):
![image](https://user-images.githubusercontent.com/93914/220300330-8c863e6c-4375-44af-b4eb-0f6cc9b471f1.png)
![image](https://user-images.githubusercontent.com/93914/220300520-c1717e8b-29cc-45bb-a041-ff420a6436b2.png)

**Comment by techmagus** [comment]
> That is because you are both changing the color and making the diacritics use bold font. The bold bit can’t be made to work, but changing the color alone is possible in Firefox (if decomposed form is used). Chrome will not change the color even if decomposed form is forced as it treats grapheme clusters as one unit for styling even if they are composed of multiple glyphs. WebKit is even worse, inline styling always breaks text layout even if the same font is used.

I updated my test-repo branch: https://github.com/techmagus/test-repo/tree/noto-diacriticals

It now includes regular/non-bold samples, and colouring diacritical marks works in Firefox, Edge, and Whale (the last two are Chromium-based). Here are screenshots of the latest update.

What I did is this:
* declare `@font-face` with different `font-family` names
* declared the new `font-family` in a class
* use that class in a diacritical mark `span`

Which then prevented Firefox, Edge, and Whale to use a precomposed glyph, since it thinks there's a different font.

(PS Unless I missed, or mixed, something, so it was giving me false results.)

![Screenshot from 2023-02-22 07-27-39](https://user-images.githubusercontent.com/1079205/220482886-2c17657c-4b95-4602-8633-b5a354a784ab.png)
![Screenshot from 2023-02-22 07-28-59](https://user-images.githubusercontent.com/1079205/220482897-953d5335-0579-49cb-8538-d4a4c25437ab.png)
![Screenshot from 2023-02-22 07-30-28](https://user-images.githubusercontent.com/1079205/220482901-f843fc62-5624-4c5b-8c9e-7ba915ac78c0.png)

**Comment by khaledhosny** [comment]
You are still using a different font for the marks (it does not matter if it is the same font file, if the browser thinks it is a different font then it will process the base and marks separately and the font data for positioning the marks will not be used).
